File: scripts/visualize/visualize-benchmark.R
#!/usr/bin/env Rscript

##############################
# libraries
##############################
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(argparse))

##############################
# command line parsing
##############################
parser <- ArgumentParser(description = "Visualize contents of a csv file containing the collected benchmark results")
parser$add_argument("-csv", "--csv-file", type = "character", nargs = 1, action="store", required = TRUE, help = "the csv file containing the results to visualize; required csv columns are: task, version, num_threads, size, run")
parser$add_argument("-col", "--column-name", type = "character", nargs = 1, action="store", required = TRUE, help = "the column name of the metric to visualize")
parser$add_argument("-metric", "--metric-name", type = "character", nargs = 1, action="store", required = TRUE, help = "the readable metric name displayed in the plot")
args <- parser$parse_args()

csv_file <- args$csv_file
column_name <- args$column_name
metric_name <- args$metric_name

##############################
# data import & transform
##############################
df <-
  read.csv(csv_file) %>%
  mutate(
    size = factor(
      size, ordered = TRUE,
      levels = c("small", "large"),
      labels = c("Small Problem Size", "Large Problem Size")
    )
  ) %>%
  group_by(task, version, num_threads, size) %>%
  mutate(
    zscore = scale(get(column_name)),
    is_normally_distributed = factor(
      shapiro.test(get(column_name))$p.value > 0.05,
      ordered = TRUE, levels = c(FALSE, TRUE)
    )
  )

num_executions <- length(unique(df$run))
task <- df$task[1]
version <- df$version[1]

##############################
# visualization
##############################
output_dir <- file.path(dirname(normalizePath(csv_file)), "plots")
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

ggplot(df, aes(x = num_threads, y = zscore, group = num_threads, colour = is_normally_distributed)) +
  facet_wrap(~ size, nrow = 2, ncol = 1) +
  geom_violin() +
  geom_boxplot(width = 0.1) +
  stat_summary(geom = "point", fun = mean, shape = 18, size = 4) +
  labs(colour = "Normally\nDistributed?") +
  xlab("#Threads") +
  ylab(str_interp("Normalized ${metric_name} (zscore)")) +
  ggtitle(str_interp("${metric_name} Deviation over ${num_executions} Executions of Task ${task} (${version})")) +
  ggpresent::theme_presentation() +
  scale_x_continuous(trans = "log2", breaks = ggpresent::trans_breaks_log2()) +
  scale_colour_manual(values = c("#E41A1C", "#000000"), labels = c("No", "Yes"))

for (dev in c("pdf", "png")) {
  ggpresent::save_presentation(path = output_dir, filename = paste(column_name, "normalized-distribution", dev, sep = "."))
}

# disable message about output grouping to not cause confusion
options(dplyr.summarise.inform = FALSE)
table_data <-
  df %>%
  summarize(
    Median = round(median(get(column_name)), digits = 2),
    Mean = round(mean(get(column_name)), digits = 2),
    SD = round(sd(get(column_name)), digits = 2)
  ) %>%
  arrange(size, num_threads) %>%
  rename(Task = task, Version = version, "#Threads" = num_threads, "Problem Size" = size)
options(dplyr.summarise.inform = TRUE)

ggplot() +
  annotation_custom(
    grob = tableGrob(
      table_data,
      rows = NULL,
      theme = ttheme_default(
        core = list(fg_params = list(cex = 1.75, hjust = 1, x = 0.975)),
        colhead = list(fg_params = list(cex = 1.75))
      )
    )
  ) +
  ggpresent::theme_presentation()

for (dev in c("pdf", "png")) {
  ggpresent::save_presentation(path = output_dir, filename = paste(column_name, "table", dev, sep = "."))
}

##############################
# cleanup
##############################
# necessary for non-interactive execution
ggpresent::cleanup_devpdf()

File: scripts/visualize/visualize-benchmark.py
import argparse
import polars as pl
import matplotlib.pyplot as plt
import os

parser = argparse.ArgumentParser(description="Visualize contents of a csv file containing the collected benchmark results")
parser.add_argument("-csv", "--csv-file", 
                    required=True, 
                    help="the csv file containing the results to visualize; required csv columns are: task, version, num_threads, size, run")
parser.add_argument("-col", "--column-name",
                    required=True,
                    help="the column name of the metric to visualize")
parser.add_argument("-m", "--metric-name",
                    required=True,
                    help="the readable metric name displayed in the plot")
parser.add_argument("-ni", "--no-ideal",
                    action="store_true",
                    help="flag indicating not to draw the ideal scaling curve")
parser.add_argument("-i", "--invert-ideal",
                    action="store_true",
                    help="invert the scaling curve for the metric so that ideal=num_threads*v[0] instead of ideal=v[0]/num_threads")
parser.add_argument("-o", "--output",
                    required=False,
                    help="Output file to write plots to (default: <folder of csv input file>/plot.png)")
parser.add_argument("-d", "--deviation",
                    action="store_true",
                    help="draw deviation intervals on data points")
parser.add_argument("-agg", "--aggregation",
                    default="avg",
                    choices=["avg", "med"],
                    help="aggregation method to use for plotting metric (default avg)")
args = parser.parse_args()

outfile = args.output or os.path.join(os.path.dirname(args.csv_file), "plot.png")

df = pl.read_csv(args.csv_file)
match args.aggregation:
    case "med":
        mode = "Median"
        df_small = df.filter(pl.col("size") == "small").group_by("num_threads").agg(pl.col(args.column_name).median()).sort("num_threads")
        df_large = df.filter(pl.col("size") == "large").group_by("num_threads").agg(pl.col(args.column_name).median()).sort("num_threads")
    case "avg":
        mode = "Average"
        df_small = df.filter(pl.col("size") == "small").group_by("num_threads").agg(pl.col(args.column_name).mean()).sort("num_threads")
        df_large = df.filter(pl.col("size") == "large").group_by("num_threads").agg(pl.col(args.column_name).mean()).sort("num_threads")

min_small = df.filter(pl.col("size") == "small").group_by("num_threads").agg(pl.col(args.column_name).min()).sort("num_threads")
max_small = df.filter(pl.col("size") == "small").group_by("num_threads").agg(pl.col(args.column_name).max()).sort("num_threads")
min_large = df.filter(pl.col("size") == "large").group_by("num_threads").agg(pl.col(args.column_name).min()).sort("num_threads")
max_large = df.filter(pl.col("size") == "large").group_by("num_threads").agg(pl.col(args.column_name).max()).sort("num_threads")

e_small = [df_small[args.column_name] - min_small[args.column_name],
           max_small[args.column_name] - df_small[args.column_name]]

e_large = [df_large[args.column_name] - min_large[args.column_name],
           max_large[args.column_name] - df_large[args.column_name]]

xt_s = [1] + [x for x in range(12, 97, 12)]
xt_l = [1] + [x for x in range(12, 97, 12)]

# TODO: how to account for errorbars on edge of plot?
xlim_s = (df_small["num_threads"][0], df_small["num_threads"][-1]) 
xlim_l = (df_large["num_threads"][0], df_large["num_threads"][-1])

fig, ax = plt.subplots(2,2, figsize=(12,10))
ax[0,0].grid()
ax[0,1].grid()
ax[1,0].grid()
ax[1,1].grid()

# Plot for small problem size
ref_small = df_small[args.column_name][0]
n0_s = df_small["num_threads"][0]

if not args.no_ideal:
    if args.invert_ideal:
        ideal_s = [ref_small * (n / n0_s) for n in df_small["num_threads"]]
    else:
        ideal_s = [ref_small / (n / n0_s) for n in df_small["num_threads"]]

    speedup_ideal_s = [n / n0_s for n in df_small["num_threads"]]

    ax[0,0].plot(df_small["num_threads"], ideal_s, color="grey")
    ax[1,0].plot(df_small["num_threads"], speedup_ideal_s, color="grey")

ax[0,0].set_title(f"Small Problem Size ({mode})")
ax[0,0].set_xlabel("Number of threads")
ax[0,0].set_ylabel(args.metric_name)
ax[0,0].set_xticks(xt_s)
ax[0,0].set_xlim(xlim_s)
if args.deviation:
    ax[0,0].errorbar(df_small["num_threads"], df_small[args.column_name], 
                     yerr=e_small, color="red", ecolor="lightcoral", capsize=2.0)
else:
    ax[0,0].plot(df_small["num_threads"], df_small[args.column_name], color="red")

# Speedup small problem size
ax[1,0].set_title("Speedup for Small Problem Size")
ax[1,0].set_xlabel("Number of threads")
ax[1,0].set_ylabel("Speedup")
ax[1,0].set_xticks(xt_s)
ax[1,0].set_xlim(xlim_s)

if args.invert_ideal:
    speedup_small = [v/ref_small for v in df_small[args.column_name]]
else:
    speedup_small = [ref_small/v for v in df_small[args.column_name]]

ax[1,0].plot(df_small["num_threads"], speedup_small, color="red")


# Plot for large problem size
ref_large = df_large[args.column_name][0]
n0_l = df_large["num_threads"][0]

if not args.no_ideal:
    if args.invert_ideal:
        ideal_l = [ref_large * (n / n0_l) for n in df_large["num_threads"]]
    else:
        ideal_l = [ref_large / (n / n0_l) for n in df_large["num_threads"]]
        
    speedup_ideal_l = [n / n0_l for n in df_large["num_threads"]]

    ax[0,1].plot(df_large["num_threads"], ideal_l, color="grey")
    ax[1,1].plot(df_large["num_threads"], speedup_ideal_l, color="grey")

ax[0,1].set_title(f"Large Problem Size ({mode})")
ax[0,1].set_xlabel("Number of threads")
ax[0,1].set_ylabel(args.metric_name)
ax[0,1].set_xticks(xt_l)
ax[0,1].set_xlim(xlim_l)
if args.deviation:
    ax[0,1].errorbar(df_large["num_threads"], df_large[args.column_name], 
                     yerr=e_large, color="red", ecolor="lightcoral", capsize=2.0)
else:
    ax[0,1].plot(df_large["num_threads"], df_large[args.column_name], color="red")

# Speedup large problem size
ax[1,1].set_title("Speedup for Large Problem Size")
ax[1,1].set_xlabel("Number of threads")
ax[1,1].set_ylabel("Speedup")
ax[1,1].set_xticks(xt_l)
ax[1,1].set_xlim(xlim_l)

if args.invert_ideal:
    speedup_large = [v/ref_large for v in df_large[args.column_name]]
else:
    speedup_large = [ref_large/v for v in df_large[args.column_name]]

ax[1,1].plot(df_large["num_threads"], speedup_large, color="red")


fig.tight_layout()
fig.savefig(outfile, dpi=200)